### PR TITLE
feat : added hashtag creation alongside with tweet create/update

### PR DESCRIPTION
**Pull request by VladislavPokhyl**
https://trello.com/c/37IwAWM6/113-add-tag-parsing-functionality-to-tweet-creation

### CODE REVIEW

File: src/services/tweets.service.ts [hunk 1]
```diff
@@ -42,7 +53,21 @@ export const updateTweet = (
...
+  const tweetOldData = await getOneById(id);
+  const updateResult = await updateOne(id, newData);
+
+  if (updateResult) {
+    const oldHashtags = TweeterUtils.extractHashtags(
+      convert(tweetOldData.content)
+    );
+    const newHashtags = TweeterUtils.extractHashtags(convert(newData.content));
+    const hashtagsDiff = newHashtags.filter((x) => !oldHashtags.includes(x));
+    if (hashtagsDiff) {
```
Comment (alexandrtovmach) [inline src/services/tweets.service.ts]:
```suggestion
    if (hashtagsDiff.length) {
```

File: src/services/tags.service.ts [hunk 2]
```diff
@@ -31,3 +32,17 @@ export const deleteTagById = async (id: string) => {
...
+    const findResult = await getByName(name);
+
+    if (findResult === null) {
+      await create({ name: name }, { initialIncrementStats: true });
+    } else {
+      await update(findResult._id, {
+        stats: { tweets: ++findResult.stats.tweets },
+      });
+    }
+  });
```
Comment (alexandrtovmach) [inline src/services/tags.service.ts]: You can optimize this algorithm:
1. Do attempt to update tag by name with new stats
2. If it's failed to update, create a new tag
3. If it's success you're done

In that way you don't need to make "find" request

File: src/repositories/tags.repository.ts [hunk 3]
```diff
@@ -1,10 +1,21 @@
...
+export const create = (
+  tagData: Partial<Tag>,
+  createOptions?: CreateOptions
+) => {
+  if (createOptions) {
+    //to be continued
+    if (createOptions?.initialIncrementStats) {
+      tagData.stats = { tweets: 1 };
+    }
+  }
```
Comment (alexandrtovmach) [inline src/repositories/tags.repository.ts]: Not sure for what do we need this "options"?

**Review by alexandrtovmach — CHANGES_REQUESTED** [review]

File: src/services/tweets.service.ts (hunk 1 above)
Comment (VladislavPokhyl) [reply src/services/tweets.service.ts]: My omission.  I'll fix it. Thanks

File: src/services/tweets.service.ts [hunk 4]
```diff
@@ -24,15 +27,23 @@ export const getTweetById = (id: string) => {
...
-export const createTweet = (
+export const createTweet = async (
   author: string,
   content: string,
   repliedTo?: string
 ) => {
-  return createOne(author, content, repliedTo);
+  const createResult = await createOne(author, content, repliedTo);
+
+  if (createResult) {
```
Comment (VitaliiDenysiuk) [inline src/services/tweets.service.ts]: what do you check here?
if error we did not come here
if object we had always true

File: src/services/tweets.service.ts [hunk 5]
```diff
@@ -42,7 +53,21 @@ export const updateTweet = (
     updatedAt: string;
   }
 ) => {
-  return updateOne(id, newData);
+  const tweetOldData = await getOneById(id);
+  const updateResult = await updateOne(id, newData);
+
+  if (updateResult) {
```
Comment (VitaliiDenysiuk) [inline src/services/tweets.service.ts]: what do you check here?
if error we did not come here
if object we had always true

File: src/services/tweets.service.ts [hunk 6]
```diff
@@ -42,7 +53,21 @@ export const updateTweet = (
...
+    const oldHashtags = TweeterUtils.extractHashtags(
+      convert(tweetOldData.content)
+    );
+    const newHashtags = TweeterUtils.extractHashtags(convert(newData.content));
+    const hashtagsDiff = newHashtags.filter((x) => !oldHashtags.includes(x));
```
Comment (VitaliiDenysiuk) [inline src/services/tweets.service.ts]: create normal name of variable.
What if we had some tags before but now delete it?
Check this situation too use function decrementStatsByName

File: src/repositories/tags.repository.ts [hunk 7]
```diff
@@ -23,3 +34,14 @@ export const update = (tagId: string, newTagData: Partial<Tag>) => {
 export const deleteById = (tagId: string) => {
   return TagModel.findByIdAndDelete(tagId);
 };
+
+export const getByName = (name: string) => {
+  return TagModel.findOne({ name: name });
+};
+
+export const incrementStatsByName = (name: string) => {
```
Comment (VitaliiDenysiuk) [inline src/repositories/tags.repository.ts]: add function decrement 
When we created tweet with tag, but after that modificated it and removed some tags from tweet

**Review by VitaliiDenysiuk — CHANGES_REQUESTED** [review]
Very important moment:
does not take into account the situation when the tag is removed when editing a tweet

File: src/services/tags.service.ts [hunk 8]
```diff
@@ -31,3 +34,31 @@ export const deleteTagById = async (id: string) => {
...
+    TweeterUtils.extractHashtags(convert(content))
+  );
+
+  if (hashTags) {
+    hashTags.forEach(async (name: string) => {
+      let result: Tag;
+      result = await incrementStatsByName(name);
+      if (!result) {
+        result = await create({ name: name }, { initialIncrementStats: true });
+      }
```
Comment (VitaliiDenysiuk) [inline src/services/tags.service.ts]:
```suggestion
      if (!(await incrementStatsByName(name))) {
        await create({ name: name }, { initialIncrementStats: true });
      }
```

**Review by VitaliiDenysiuk — CHANGES_REQUESTED** [review]

File: src/repositories/tags.repository.ts (hunk 3 above)
Comment (alexandrtovmach) [reply src/repositories/tags.repository.ts]: make default stats tweet for Tag model as 1, and drop this logic

**Review by alexandrtovmach — APPROVED** [review]

**Review by DavidKorystov — APPROVED** [review]

**Pull request merged**